Documentation/design-docs/AssemblyLoadContext.Activate.md

```diff
@@ -0,0 +1,306 @@
+* Executing methods defined within an assembly loaded by the `AssemblyLoadContext`
+
+There is a need for a general solution to allow code executing outside a given `AssemblyLoadConxt` to behave as if it was loaded within the context.
+
+This is in part because of the design of `AssemblyLoadContext` which
+* Requires `System.Private.CoreLib` to be loaded exclusively in the `Default` context.
+* Strongly recommends the runtime framework be completely loaded into the `Default` context.
+* Recommends most frameworks be loaded exclusively in the `Default` context.
+
+This proposal is for a mechanism for code to explicitly set a specific `AssemblyLoadContext` as the active context.
+
+## Proposed API extension(s)
+```C#
+namespace System.Runtime
+{
+  partial class AssemblyLoadContext
+  {
+    // When set supersedes implicit context as determine by callers assembly.
+    // Affects most/all APIs which are (or will be) marked [System.Security.DynamicSecurityMethod]
+    // Implemented as AsyncLocal<AssemblyLoadContext> to allow setting to follow async & threads
+    public static AssemblyLoadContext ActiveContext { get { return _asyncLocalActiveContext.Value; }}
+
+    // Allow client to explicitly set the active context for a using block
+    public IDisposable Activate();
+
+    // Convenience method to call AssemblyLoadContext on an assemblies parent
+    // Activate(null) will explicitly clear the ActiveContext until the IDisposable.Dispose()
+    static public IDisposable Activate(Assembly activating);
+  }
+}
+```
+
+`ActiveContext` is stored in `AsyncLocal<AssemblyLoadContext>`. It represents the most recent `Activated()` `AssemblyLoadContext`.
+
+
+`Activate()` is intended to be used in a using block and/or statement.
+
+When `Activate()` is called it make the activated `AssemblyLoadContext` the current executing assembly.  This will impact the preference order for loading assemblies and resolving type names.
+
+Before `Activate()` is called, the behavior of the exitisting API's will remanin unchanged.  The `AssemblyLoadContext` containing the API caller will determine the preferred `AssemblyLaodContext`.
+
+Once `Activate()` is called, the associated `AssemblyLoadContext` will become the preferred `AssemblyLoadContext`.  The value will be reflected in `ActiveContext`
+* This preference will affect the current thread and its asynchronous control flow.
+* The behavior will persist in the current thread until the returned `IDisposable` is disposed.
+* **or**  until a nested call supercedes the preference.
+* `IDisposable.Dispose()` will restore the previous `ActiveContext` to support nesting.
+* Async workers will inherit their parent's ActiveContext when they are created. Their behavior will not be affected when `IDisposable.Dispose()` is called in the parent thread.  The worker can call `Activate()` to modifiy its own `ActiveContext`.
+* `Thread`s and `ThreadPool` queued work items should also be affected
```

Comment:
Okay, so the activation context will have thread affinity and not be process wide, right? If it was process wide there are potentially a lot of scenarios where plug-ins could do bad things and play heck with the stability of the app - think Visual Studio. On the other hand if the context has thread affinity then we can get into scenarios where objects may have thread affinity, which is one of the lessons from COM. We can all enumerate the negatives of both models, but there are good reasons for both as well. That aside, I don't know how either really fits into the .NET model.

Reply:
@AaronRobinsonMSFT I used the should language because it expresses my wish. I will need to prototype to understand the natural behavior of `AsyncLocal<T>` to see what is achievable.
We are trying to model a behavior where a given thread enters a `AssemblyLoadContext` like it would an `AppDomain`.  While in the `AssemblyLoadContext` all framework code is presumed to be acting on behalf of the `AssemblyLoadContext`.  Therefore any spawned async control flow including threads would automatically inherit the parent threads behavior.

Reply:
>Therefore any spawned async control flow including threads would automatically inherit the parent threads behavior.

Yep. I get that and it is a laudable goal. My critique is that the ALC is now a transient concept that is both related to Assembly _and_ thread. This overloaded scope is going to be very confusing. For example, if my assembly was loaded in ALC1 but I call a function on that assembly that dynamically loads another assembly in a delayed fashion, but the thread is operating under ACL2, the logic for ALC2 can become inadvertently dependent on ALC1 which may or may not be entirely understood/expected by the caller. I expect you know this, but my question is how should I expect this to work in a world where the author of the assembly expects the same ALC that loaded itself to be used for all its dependencies.